app/src/main/java/com/codepath/apps/critterfinder/fragments/PetDetailsFragment.java

```diff
@@ -49,20 +65,109 @@ public void onCreate(Bundle savedInstanceState) {
+    }
+
+    @Override
+    public void onAttach(Context context) {
+        super.onAttach(context);
+        if (context instanceof OnDetailsFragmentListener) {
+            mDetailsFragmentListener = (OnDetailsFragmentListener)context;
+        } else {
+            throw new RuntimeException(context.toString());
+        }
     }
 
     private void setupPetDetailsView() {
-        mPetGender.setText(mPet.getSizeSexAge());
+        mPetInfo.setText(mPet.getSizeSexAge());
         mPetBreed.setText(mPet.getBreedFullName());
         mPetDescription.setText(mPet.getDescription());
         mShelterContactName.setText(mPet.getContactName());
         mShelterContactPhone.setText(mPet.getContactPhone());
-        mShelterContactState.setText(mPet.getContactState());
-        mShelterCity.setText(mPet.getContactCity()+", ");
+        mShelterLocation.setText(mPet.getLocation());
         mShelterEmail.setText(mPet.getContactEmail());
     }
+
+    // open in map
+    public void onLocate() {
+        String address = mPet.getLocation();
+        // Looks like the address is always a PO Box # which won't help us to map the location !
+//        if (mPet.getContactAdress2().length() > 0) {
+//            address = mPet.getContactAdress1();
+//            address += " " + mPet.getLocation();
+//        } else
+//        {
+//            address = mPet.getLocation();
+//        }
+        String uriEncoded;
+        try {
+            uriEncoded = URLEncoder.encode(address, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+            return;
+        }
+        String uri = String.format(Locale.ENGLISH, "geo:0,0?q=%s", uriEncoded);
+        Log.d("PetDetailsFragment", "Opening address in Google Maps: " + uri);
+
+        if (mDetailsFragmentListener != null) {
+            mDetailsFragmentListener.onMap(uri);
+        }
+    }
+
+    // Call the shelter
+    public void onCall() {
+        Intent intent = new Intent(Intent.ACTION_CALL, Uri.parse("tel:" + mPet.getContactPhone()));
+        startActivity(intent);
+    }
+
+    // Create an Email to the shelter
+    public void onEmail() {
+        Intent intent = new Intent(Intent.ACTION_SEND);
+        intent.setType("plain/text");
+
+        final String subject = getString(R.string.contact_pet_subject, mPet.getName());
+        final String body = getString(R.string.contact_pet_body, mPet.getName());
+
+        intent.putExtra(Intent.EXTRA_EMAIL, new String[]{mPet.getContactEmail()});
+        intent.putExtra(Intent.EXTRA_SUBJECT, subject);
+        intent.putExtra(Intent.EXTRA_TEXT, body);
+
+        try {
+            startActivityForResult(intent, 0);
+        } catch (ActivityNotFoundException e) {
+            Snackbar.make(mFragmentView, getString(R.string.contact_no_email_client), Snackbar.LENGTH_LONG).show();
+        }
+    }
+
+    public interface OnDetailsFragmentListener {
+        void onMap(String url);
+    }
```

Comment:
good stuff. Just sharing a tip for the future - I think this public interface could be extended to also have a onCall, and onEmail notification. I read in the documentation that communication to launch another activity or communicate with another activity should get delegated out to the fragment's parent activity like you did here with onMap.. So the parent activity is responsible for all launching and communication with other activities.